torch/distributed/fsdp/fully_sharded_data_parallel.py

```diff
@@ -2273,6 +2279,107 @@ def shard_full_optim_state_dict(
         }
         return sharded_optim_state_dict
 
+    @staticmethod
```

Comment:
This functionality is actually not unique to `FSDP`. It _could_ be used generally for PyTorch. However, nowhere else have we seen keying by parameter name, so I have put this inside `fully_sharded_data_parallel.py` for now.

Reply:
sounds good!